File: 1423-maximum-points-you-can-obtain-from-cards/1423-maximum-points-you-can-obtain-from-cards.cpp
class Solution {
public:
    int maxScore(vector<int>& v, int k) {
        int N = v.size();
        if (k >= N)
            return accumulate(v.begin(), v.end(), 0);
        vector<int> pre(N + 1, 0), suff(N + 2, 0);
        
        pre[1] = v[0];
        for (int i = 2; i <= N; i++) pre[i] = pre[i - 1] + v[i - 1];
        
        suff[N] = v[N - 1];
        for (int i = N - 1; i >= 1; i--) suff[i] = suff[i + 1] + v[i - 1];
        
        int ans = 0;
        for (int i = 0; i <= k; i++) {
            int left = pre[i];
            int right = suff[N - (k - i) + 1];
            ans = max(ans, left + right);
        }
        
        return ans;
    }
};

File: 1423-maximum-points-you-can-obtain-from-cards/1423-maximum-points-you-can-obtain-from-cards.py
class Solution:
    def maxScore(self, a: List[int], k: int) -> int:
        if k == len(a):
            return sum(a)
        
        N = len(a)
        pre = (N + 2) * [0]
        suff = (N + 2) * [0]
        
        pre[1] = a[0]
        for i in range(2, N + 1) :
            pre[i] = pre[i - 1] + a[i - 1]
            
        suff[N] = a[N - 1]
        for i in range(N - 1, 0, -1) :
            suff[i] = suff[i + 1] + a[i - 1]
        
        ans = 0
        for i in range(k + 1) :
            left, right = pre[i], suff[N - (k - i) + 1]
            ans = max(ans, left + right)
            
        return ans